Python/_warnings.c

```diff
@@ -257,7 +258,9 @@ get_once_registry(PyInterpreterState *interp)
         Py_DECREF(registry);
         return NULL;
     }
+    Py_BEGIN_CRITICAL_SECTION(registry);
```

Comment:
We need locking around both the reads (i.e., `return st->once_registry` above) and the writes. Additionally, we should be locking something like the module state (`st`), instead of the `registry` value. Otherwise, threads that set `st->once_registry` to different values will be locking different `registry` objects.

It might be simplest to do the locking around `warn_explicit` like you suggested in the issue.

Reply:
Ok, thanks for the pointers.

Reply:
Addressed in d222f7c83ed1dfe53b47eabce9e0ccaefdb6a83c